File: flow_sdk/builtin/journey_journal.py
"""JourneyJournal — a user's private progress through a Journey.

**The journal IS the progress object** — every `Journey` method (`launch`,
`restart`, `advance`, `progress`, `history`, `resume`) returns this row; there is
no separate progress DTO. It is the DURABLE cursor: the live flow run is only the
advance mechanism / one-liner and dies on server restart, while the journal
survives and re-parks the run.

Many journals may exist per (user, journey) — restart archives the current one and
starts a fresh run — but **at most one is active** (status ∈ {new, launched}).
Progress lives HERE, never in the shared Journey folder, so collaborators sharing
one authored Journey keep independent progress.

``steps_left`` is the badge count. ``entries`` is an append-only ledger of step
outcomes. The step DESCRIPTORS are not stored here — they are read from the
journey's ``graph.json``; the UI derives each step's done/current/upcoming state
from ``cursor`` + ``entries``.
"""
from __future__ import annotations

from enum import Enum
from typing import Any, ClassVar

from flow_sdk.api.api_types.api_field import APIField
from flow_sdk.core import Entity
from flow_sdk.schema.types import EntityType


class JourneyStatus(str, Enum):
    """new → launched → complete, plus `restarted` for a superseded journal."""

    NEW = "new"              # launched, cursor at entry, nothing advanced yet
    LAUNCHED = "launched"    # in progress (≥1 step advanced)
    COMPLETE = "complete"    # every step done
    RESTARTED = "restarted"  # abandoned by a restart, or superseded by a resume


#: Non-terminal statuses. The one-active invariant: at most one journal per
#: (user, journey) may carry one of these at any time.
ACTIVE_STATUSES: frozenset[str] = frozenset({JourneyStatus.NEW.value, JourneyStatus.LAUNCHED.value})


class JourneyJournal(Entity):
    type: str = APIField(default=EntityType.JOURNEY_JOURNAL.value)

    journey_id: str = APIField(default="")
    user_id: str = APIField(default="")
    status: str = APIField(default=JourneyStatus.NEW.value,
                           description="new | launched | complete | restarted")
    run_id: str = APIField(default="", description="Live AgenticFlowRun advancing this journal.")
    cursor: str = APIField(default="", description="Current guided_step node id ('' when complete).")
    total_steps: int = APIField(default=0)
    steps_left: int = APIField(default=0, description="Badge count = total_steps - completed.")
    entries: list[dict[str, Any]] = APIField(
        default_factory=list, description="Append-only step outcomes: {node_id, event, at}."
    )

    _api_visible: ClassVar[bool] = True

    @property
    def is_active(self) -> bool:
        return self.status in ACTIVE_STATUSES

File: ts_sdk/src/entities/journey/journey-journal.ts
import { APIEntity, registerEntity } from '../../APIEntity';
import { IEntity } from '../../IEntity';
import type { JourneyGraph } from './journey-graph';

/** new → launched → complete, plus `restarted` for a superseded journal. */
export type JourneyStatus = 'new' | 'launched' | 'complete' | 'restarted';

/** One recorded step transition inside a journal. */
export interface JourneyJournalEntry {
  node_id: string;
  event: string;
  at?: string;
}

export interface IJourneyJournal extends IEntity {
  journey_id?: string;
  user_id?: string;
  status?: JourneyStatus;
  run_id?: string;
  cursor?: string;
  total_steps?: number;
  steps_left?: number;
  entries?: JourneyJournalEntry[];
}

// `implements IJourneyJournal` only checks the class; it contributes no members, so every
// field declared solely on IJourneyJournal read as "does not exist". deepAssign populates
// them from the wire — this merge makes them part of the class type.
// eslint-disable-next-line @typescript-eslint/no-empty-object-type
export interface JourneyJournal extends Omit<IJourneyJournal, 'expand' | 'id' | 'is_private' | 'members'> {}

/**
 * Per-user progress through a `Journey` — and THE object every journey
 * method returns (there is no separate progress DTO). `cursor` is the current
 * step node id; `steps_left` is the badge count. The step DESCRIPTORS are not
 * here: they come from the journey's {@link JourneyGraph}, and each step's
 * done/current/upcoming state is derived from `cursor` + `entries` by the
 * accessors below.
 *
 * The backend is the single writer — the frontend only reads this.
 */
@registerEntity
export class JourneyJournal extends APIEntity<JourneyJournal> implements IJourneyJournal {
  static type: string = 'journey_journal';
  journey_id?: string;
  user_id?: string;
  status?: JourneyStatus;
  run_id?: string;
  cursor?: string;
  total_steps?: number;
  steps_left?: number;
  entries?: JourneyJournalEntry[];

  constructor(entity: Partial<IJourneyJournal> = {}) {
    super(entity);
    this.journey_id = entity.journey_id;
    this.user_id = entity.user_id;
    this.status = entity.status;
    this.run_id = entity.run_id;
    this.cursor = entity.cursor;
    this.total_steps = entity.total_steps;
    this.steps_left = entity.steps_left;
    this.entries = entity.entries;
  }

  /** True while this journal is the one being worked on (`new` | `launched`). */
  get isActive(): boolean {
    return this.status === 'new' || this.status === 'launched';
  }

  get isComplete(): boolean {
    return this.status === 'complete';
  }

  /** Where the cursor sits in `graph`, or -1 when it names no known step
   *  (a journal from an older revision of the graph). */
  indexIn(graph: JourneyGraph): number {
    return graph.indexOf(this.cursor);
  }

  /** Node ids with a recorded outcome — done OR skipped. Both are "behind you",
   *  which is what every tick-mark in the tray and viewer means. */
  doneNodeIds(): Set<string> {
    return new Set((this.entries ?? []).map((e) => e.node_id));
  }

  /**
   * THE journal-selection policy: the active journal wins; otherwise the most
   * recently touched one.
   *
   * Never an arbitrary row — an archived `restarted` journal still carries a
   * cursor, and picking it would make the manager re-present a stale step.
   * This lived as a `useMemo` inside a React hook, which meant the rule could
   * not be stated or tested anywhere else even though the badge, the viewer and
   * the tray all depend on it agreeing.
   */
  static pick(journals: readonly JourneyJournal[], journeyId: string | null | undefined): JourneyJournal | null {
    if (!journeyId) return null;
    const mine = journals.filter((j) => j.journey_id === journeyId);
    const active = mine.find((j) => j.isActive);
    if (active) return active;
    // `APIEntity.compare` is the shipped comparator (same call shape as
    // `Project.compare('updated_date')` in use-projects) — and it compares the
    // raw ISO strings rather than parsing, which is well-defined where a
    // Date.parse subtraction would go NaN on a missing date.
    return [...mine].sort(JourneyJournal.compare('updated_date', 'desc'))[0] ?? null;
  }
}
